dvc/api.py

```diff
@@ -120,94 +127,134 @@ def _make_repo(repo_url=None, rev=None):
+        object, named object specification and resolved paths to deps.
+        """
+        with _make_repo(repo, rev=rev) as _repo:
+            _require_dvc(_repo)
+            try:
+                yield SummonDesc(_repo, summon_file)
+            except SummonError as exc:
+                raise SummonError(
+                    str(exc) + " at '{}' in '{}'".format(summon_file, _repo)
+                ) from exc.__cause__
+
+    def deps_paths(self, dobj):
+        return dobj["summon"].get("deps", [])
+
+    def deps_abs_paths(self, dobj):
+        return [
+            os.path.join(self.repo.root_dir, p) for p in self.deps_paths(dobj)
+        ]
 
-    def _pull_deps(self):
-        if not self._deps:
-            return
+    def outs(self, dobj):
+        return [
+            self.repo.find_out_by_relpath(d) for d in self.deps_paths(dobj)
+        ]
 
-        outs = [self.repo.find_out_by_relpath(d) for d in self._deps]
+    def pull(self, dobj):
+        outs = self.outs(dobj)
 
         with self.repo.state:
             for out in outs:
                 self.repo.cloud.pull(out.get_used_cache())
                 out.checkout()
 
+    def push(self, dobj):
+        paths = self.deps_abs_paths(dobj)
 
-def _get_object_spec(name, path):
-    """
-    Given a summonable object's name, search for it on the given file
-    and return its description.
-    """
-    try:
-        with builtin_open(path, "r") as fobj:
-            content = SUMMON_FILE_SCHEMA(ruamel.yaml.safe_load(fobj.read()))
-            objects = [x for x in content["objects"] if x["name"] == name]
+        with self.repo.state:
+            for path in paths:
+                self.repo.add(path)
+                self.repo.add(path)
+
+    def get_dobject(self, name, default=False):
+        """
+        Given a summonable object's name, search for it on the given content
+        and return its description.
+        """
+        objects = [x for x in self.content[DOBJ_SECTION] if x["name"] == name]
 
         if not objects:
-            raise SummonError("No object with name '{}'".format(name))
+            raise SummonErrorNoObjectFound(
+                "No object with name '{}'".format(name)
+            )
         elif len(objects) >= 2:
             raise SummonError(
                 "More than one object with name '{}'".format(name)
             )
 
         return objects[0]
 
-    except FileNotFoundError as exc:
-        raise SummonError("Summon file not found") from exc
-    except ruamel.yaml.YAMLError as exc:
-        raise SummonError("Failed to parse summon file") from exc
-    except Invalid as exc:
-        raise SummonError(str(exc)) from exc
+    def update_dobj(self, new_dobj, overwrite=False):
+        try:
+            name = new_dobj["name"]
+            dobj = self.get_dobject(name)
+
+            if overwrite:
+                idx = self.content[DOBJ_SECTION].index(dobj)
+                self.content[DOBJ_SECTION][idx] = new_dobj
```

Comment:
Maybe we should use dict with name as key instead of a list.

Reply:
Gush... great idea.